src/category_theory/connected.lean

```diff
@@ -0,0 +1,374 @@
+/-
+Copyright (c) 2020 Bhavik Mehta. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bhavik Mehta
+-/
+
+import category_theory.limits.shapes.pullbacks
+import category_theory.limits.shapes.binary_products
+import category_theory.limits.shapes.equalizers
+import category_theory.limits.preserves
+
+/-!
+# Connected category
+
+Define a connected category
+-/
+
+universes v₁ v₂ u₁ u₂
+
+open category_theory category_theory.category category_theory.limits
+namespace category_theory
+
+section connected
+-- See note [default priority]
+set_option default_priority 100
+/--
+We define a connected category as a _nonempty_ category for which every
+functor to a discrete category is constant.
+
+NB. Some authors include the empty category as connected, we do not.
+We instead are interested in categories with exactly one 'connected
+component'.
+
+This allows us to show that the functor X ⨯ - preserves connected limits.
+-/
+class connected (J : Type v₂) [𝒥 : category.{v₁} J] extends inhabited J :=
+(iso_constant : Π {α : Type v₂} (F : J ⥤ discrete α), F ≅ (functor.const J).obj (F.obj default))
+end connected
+
+section J
+variables {J : Type v₂} [𝒥 : category.{v₁} J]
+include 𝒥
+
+/-- If J is connected, any functor to a discrete category is constant on the nose. -/
+lemma any_functor_eq_constant [conn : connected J] {α : Type v₂} (F : J ⥤ discrete α) :
+  F = (functor.const J).obj (F.obj (default J)) :=
```

Comment:
As usual, I dislike equations between functors! Can we just directly prove `F.obj j = F.obj (default j)` and `F.map f = eq.to_hom ... \gg eq_to_hom ...`?

Reply:
It turned out I only needed the first anyway so I just proved the first instead of this. The second is true by subsingleton.elim anyway.